### ISSUE TITLE
Git: Basic Merge functionality

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#637]
I would like to have an option to _merge_ branches right from Moonshine menu-options - which now I require to do by going to the Terminal or use any 3rd party software. 

The _merge_ functionality need not to be detailed (at least in first phase of development), but provide option to choose which branch to merge, run merge command and display its output to the console.

**Comment by rat-moonshine** [comment]
Having the basic merge functionality added, any conflict or other error now reports to console only, therefore user requires to fix the conflicts by themselves. A typical conflict reports in the console as they return by Git:

```
: Initiating the merge process...
: Auto-merging README.md
: CONFLICT (content): Merge conflict in README.md
: Automatic merge failed; fix conflicts and then commit the result.
: ...process completed
```

Perhaps we can create a window contains a list of conflict files (if Git returns it that way) which can help the user to read more clearly and open to the editor if requires. @JoelProminic do you have any suggestions on this (?)

**Comment by JoelProminic** [comment]
For detecting conflicts, you can use the command from [this page](https://stackoverflow.com/questions/3065650/whats-the-simplest-way-to-list-conflicted-files-in-git):

    git diff --name-only --diff-filter=U 

This will list only the files with unmerged changes.

A simple but helpful solution we could start with would be:
- Give the user an alert if there are conflicts after a merge or pull
- Show a list of the unmerged files in a new console tab
- Allow the user to click on a file in this list to open the file and manually resolve the conflict


Here are some ideas for more advanced solutions or features.  We may want to keep these for a future improvements:
- Show a split screen showing the local and remote differences, for easy comparison
- Show a GUI diff of the remote and local file.  A GUI diff tool is something that we could use elsewhere in the application
- Mark the conflicts in the project tree rather than in a console tab.  This would probably be better as a feature showing all of the local Git/SVN modifications to the project files.

**Comment by rat-moonshine** [comment]
I added two functional buttons to the conflict-view for now:

![image](https://user-images.githubusercontent.com/24370870/75774235-89a3ed80-5d75-11ea-9a7a-c6f303e3c6c3.png)

When 'Open file(s)' are suppose to open the selected file(s) to the editor, 'Mark Resolve' simply removes the items from the list. I'm not sure how useful the latter one though, please advise.

**Comment by JoelProminic** [comment]
My expectation for "Mark Resolve" (should be "Mark as Resolved") is that this would no longer show up as a conflict in "git status".   The command for this is "git add <file>.

However, this raises some questions about how we are handling staging with Moonshine.   Currently, staging (git add) is only done in the Commit window, based on which files are selected.   If we "git add" a file outside of this interface, that does not affect the the checkboxees in the Commit prompt, but I think it will be included in the commit anyway.

I discussed some other concerns about Git staging and reverts with @rat-moonshine today, but I don't have time to do a proper writeup for this today.  We can create an issue for this, and consider  how to handle it in a future release.  Some reminders:
- Allow the user to unstage (git reset) a staged file based on the selection in the Commit window
- Currently "Revert Selected" in the Status window runs a "git reset" for new files, but this won't actually clear the file from the Status window

**Comment by rat-moonshine** [comment]
- Double-clicking of conflict item now opens to editor
- 'git add' perform against items selected for Mark as Resolved action

We need to auto-select files those already staged through this action when opening the Commit window, but this can be a task for issue #682 .

**Comment by JoelProminic** [comment]
I found that I could not revert merged files using `git checkout -- <file>`.  This command does nothing on the command line for this case.  Instead, I have to use `git merge --abort`, which reverts all merge changes.  By my understanding, this happens because git is using MERGE_HEAD instead of HEAD after a `git merge` command.  I found most of this information  [here](https://stackoverflow.com/questions/101752/i-ran-into-a-merge-conflict-how-can-i-abort-the-merge).

**Comment by JoelProminic** [comment]
The updates worked like I expected.  However, the "Mark Resolve" button should be "Mark as Resolved".

**Comment by rat-moonshine** [comment]
I fixed type-error while "Mark as Resolved" command. 

However, I see some strange listing, this is probably after some of my today's changes. I will fix them.
![image](https://user-images.githubusercontent.com/24370870/77938972-205fbd80-72d4-11ea-8534-722f27e0a7f8.png)

**Comment by rat-moonshine** [comment]
It turned out that it is more convenient to use the same diff-check command that we use for Commit/Status windows to parse conflict files (`git status --porcelain`), rather than constructing a different command to get conflict files list (`git  diff --name-only --diff-filter=U`). This change will give us more control over the collection usage across different sections.

Having the changes in the conflict files list doesn't return anymore unexpected [messages](https://github.com/prominic/Moonshine-IDE/issues/637#issuecomment-606113314). Conflict window now works properly.

**Comment by piotrzarzycki21** [comment]
@Aszusz can you talk with @rat-moonshine and test it ? I would like to close it this week.

**Comment by rat-moonshine** [comment]
The current implementation is not the final version of what we may want on git-merge and related functionalities. We still need to chawk-out some plan on what we're discussing on #682 .

**Comment by JoelProminic** [comment]
Yeah, this issue has been stuck for a while because it could be a major refactoring for the code.  We can bump this to the next milestone if we want to get a release out soon.
